pkg/didcomm/protocol/messagepickup/service.go

```diff
@@ -0,0 +1,612 @@
+		ID:       msg.ID(),
+		Messages: msgs,
+	}
+
+	return s.outbound.SendToDID(batch, myDID, theirDID)
+}
+
+func (s *Service) handleBatch(msg service.DIDCommMsg) error {
+	// unmarshal the payload
+	batchMsg := &Batch{}
+
+	err := msg.Decode(batchMsg)
+	if err != nil {
+		return fmt.Errorf("batch message unmarshal : %v", err)
+	}
+
+	// check if there are any channels registered for the message ID
+	batchCh := s.getBatchCh(batchMsg.ID)
+
+	if batchCh != nil {
+		// invoke the channel for the incoming message
+		batchCh <- *batchMsg
+	}
+
+	return nil
+}
+
+func (s *Service) handleNoop(msg service.DIDCommMsg) error {
+	// unmarshal the payload
+	request := &Noop{}
+
+	err := msg.Decode(request)
+	if err != nil {
+		return fmt.Errorf("noop message unmarshal : %v", err)
+	}
+
+	return nil
+}
+
+type inbox struct {
+	DID               string          `json:"DID"`
+	MessageCount      int             `json:"message_count"`
+	LastAddedTime     time.Time       `json:"last_added_time,omitempty"`
+	LastDeliveredTime time.Time       `json:"last_delivered_time,omitempty"`
+	LastRemovedTime   time.Time       `json:"last_removed_time,omitempty"`
+	TotalSize         int             `json:"total_size,omitempty"`
+	Messages          json.RawMessage `json:"messages"`
+}
+
+// DecodeMessages Messages
+func (r *inbox) DecodeMessages() ([]*Message, error) {
+	var out []*Message
+	err := json.Unmarshal(r.Messages, &out)
+
+	if err != nil {
+		out = []*Message{}
+	}
+
+	return out, nil
+}
+
+// EncodeMessages Messages
+func (r *inbox) EncodeMessages(msg []*Message) error {
+	d, err := json.Marshal(msg)
+	if err != nil {
+		return fmt.Errorf("unable to marshal: %v", err)
+	}
+
+	r.Messages = d
+	r.MessageCount = len(msg)
+	r.TotalSize = len(d)
+
+	return nil
+}
+
+// AddMessage add message to inbox
+func (s *Service) AddMessage(message *model.Envelope, theirDID string) error {
+	outbox, err := s.getInbox(theirDID)
+	if err != nil {
+		return fmt.Errorf("unable to pull messages: %v", err)
+	}
+
+	msgs, err := outbox.DecodeMessages()
+	if err != nil {
+		return fmt.Errorf("unable to decode messages: %v", err)
+	}
+
+	m := Message{
+		ID:        uuid.New().String(),
+		AddedTime: time.Now(),
+		Message:   message,
+	}
+
+	msgs = append(msgs, &m)
+
+	outbox.LastDeliveredTime = time.Now()
+	outbox.LastRemovedTime = time.Now()
+
+	err = outbox.EncodeMessages(msgs)
+	if err != nil {
+		return fmt.Errorf("unable to pull messages: %v", err)
+	}
+
+	err = s.putInbox(theirDID, outbox)
+	if err != nil {
+		return fmt.Errorf("unable to pull messages: %v", err)
+	}
+
+	return nil
+}
+
+func (s *Service) pullMessages(theirDID string, size int) ([]*Message, error) {
+	outbox, err := s.getInbox(theirDID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to pull messages: %v", err)
+	}
+
+	msgs, err := outbox.DecodeMessages()
+	if err != nil {
+		return nil, fmt.Errorf("unable to pull messages: %v", err)
+	}
+
+	c := len(msgs)
+	end := int(math.Min(float64(size), float64(c)))
+
```

Comment:
Please remove it.

Reply:
Are you saying `end := int(math.Min(float64(size), float64(len(msgs))))` is preferred, or do you not like the math dance in general?

Reply:
Sorry I got confused. I see what you doing.
nit: it would be more readable 
```
	end := len(msgs)
	if size <end {
		end = size
	}
```

The lock/unlock work incorrectly. You're locking the database instead of your logic. The database is thread-safe. 
Locks in `getInbox` and `putInbox` don’t  make sense.

Reply:
I think they allow thread-safe operation within putInbox()  and getInbox() respectively, but should be refactored such that thread-safety is guaranteed *across* putInbox() and getInbox() (eg. in `AddMessage()`).

Reply:
I think the easiest way would be to have a queue of operations and 1 go routine processing them. Each protocol message will have its own operation.